webapp/src/Controller/Jury/ShadowDifferencesController.php

```diff
@@ -0,0 +1,207 @@
+     */
+    protected $em;
+
+    public function __construct(
+        DOMJudgeService $dj,
+        SubmissionService $submissions,
+        SessionInterface $session,
+        EntityManagerInterface $em
+    ) {
+        $this->dj          = $dj;
+        $this->submissions = $submissions;
+        $this->session     = $session;
+        $this->em          = $em;
+    }
+
+    /**
+     * @Route("", name="jury_shadow_differences")
+     */
+    public function indexAction(Request $request)
+    {
+        $shadowMode = DOMJudgeService::DATA_SOURCE_CONFIGURATION_AND_LIVE_EXTERNAL;
+        $dataSource = $this->dj->dbconfig_get('data_source', DOMJudgeService::DATA_SOURCE_LOCAL);
+        if ($dataSource != $shadowMode) {
+            $this->addFlash('danger', sprintf(
+                'Shadow differences only supported when data_source is %d',
+                $shadowMode
+            ));
+            return $this->redirectToRoute('jury_index');
+        }
+
+        // Close the session, as this might take a while and we don't need the session below
+        $this->session->save();
+
+        $contest        = $this->dj->getCurrentContest();
+        $verdictsConfig = $this->dj->getDomjudgeEtcDir() . '/verdicts.php';
+        $verdicts       = array_merge(['judging' => 'JU'], include $verdictsConfig);
+
+        $used         = [];
+        $verdictTable = [];
+        // pre-fill $verdictTable to get a consistent ordering
+        foreach ($verdicts as $verdict => $abbrev) {
+            foreach ($verdicts as $verdict2 => $abbrev2) {
+                $verdictTable[$verdict][$verdict2] = [];
+            }
+        }
+
+        /** @var Submission[] $submissions */
+        $submissions = $this->em->createQueryBuilder()
+            ->from(Submission::class, 's', 's.submitid')
+            ->leftJoin('s.external_judgements', 'ej', Join::WITH, 'ej.valid = 1')
+            ->leftJoin('s.judgings', 'j', Join::WITH, 'j.valid = 1')
+            ->select('s', 'ej', 'j')
+            ->andWhere('s.contest = :contest')
+            ->setParameter(':contest', $contest)
+            ->getQuery()
+            ->getResult();
+
+        // Helper function to add verdicts
+        $addVerdict = function ($unknownVerdict) use ($verdicts, &$verdictTable) {
+            // add column to existing rows
+            foreach ($verdicts as $verdict => $abbreviation) {
+                $verdictTable[$verdict][$unknownVerdict] = [];
+            }
+            // add verdict to known verdicts
+            $verdicts[$unknownVerdict] = $unknownVerdict;
+            // add row
+            $verdictTable[$unknownVerdict] = [];
+            foreach ($verdicts as $verdict => $abbreviation) {
+                $verdictTable[$unknownVerdict][$verdict] = [];
+            }
+        };
+
+        // Build up the verdict matrix
+        foreach ($submissions as $submitid => $submission) {
+            /** @var ExternalJudgement|null $externalJudgement */
```

Comment:
More a nit: why is the one table name ...judgement and the other one ...judging?

Reply:
I picked judgement because that is how CLICS calls it.

Reply:
ACK.